File: application/app/src/main/java/com/yahyeet/boardbook/model/handler/populator/UserPopulator.java
package com.yahyeet.boardbook.model.handler.populator;

import com.yahyeet.boardbook.model.entity.Match;
import com.yahyeet.boardbook.model.entity.MatchPlayer;
import com.yahyeet.boardbook.model.entity.User;
import com.yahyeet.boardbook.model.repository.IMatchPlayerRepository;
import com.yahyeet.boardbook.model.repository.IMatchRepository;
import com.yahyeet.boardbook.model.repository.IUserRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

public class UserPopulator {
	private IMatchPlayerRepository matchPlayerRepository;
	private IMatchRepository matchRepository;
	private IUserRepository userRepository;

	public UserPopulator(IMatchPlayerRepository matchPlayerRepository, IMatchRepository matchRepository, IUserRepository userRepository) {
		this.matchPlayerRepository = matchPlayerRepository;
		this.matchRepository = matchRepository;
		this.userRepository = userRepository;
	}

	public CompletableFuture<User> populate(User user, Map<String, Boolean> config) {
		User populatedUser = new User(user.getName());
		populatedUser.setId(user.getId());

		Boolean shouldFetchFriends = config.getOrDefault("friends", false);
		Boolean shouldFetchMatches = config.getOrDefault("matches", false);

		CompletableFuture<List<User>> friendsFuture = CompletableFuture.completedFuture(new ArrayList<>());
		if (shouldFetchFriends != null && shouldFetchFriends) {
			friendsFuture = userRepository.findFriendsByUserId(populatedUser.getId());
		}

		CompletableFuture<List<MatchPlayer>> futureMatchPlayers = CompletableFuture.completedFuture(new ArrayList<>());
		if (shouldFetchMatches != null && shouldFetchMatches) {
			futureMatchPlayers = matchPlayerRepository.findMatchPlayersByUserId(populatedUser.getId());
		}

		CompletableFuture<List<Match>> futureMatches = futureMatchPlayers.thenCompose(matchPlayers -> {
			List<CompletableFuture<Match>> allMatchFutures =
				matchPlayers
					.stream()
					.map(matchPlayer -> matchRepository.find(matchPlayer.getMatch().getId()))
					.collect(Collectors.toList());

			CompletableFuture<Void> allOfFutureAllMatchFutures = CompletableFuture.allOf(allMatchFutures.toArray(new CompletableFuture[0]));
			return allOfFutureAllMatchFutures.thenApply(
				future ->
					allMatchFutures
						.stream()
						.map(CompletableFuture::join)
						.collect(Collectors.toList())
			);
		});

		return friendsFuture.thenCombine(futureMatches, (friends, matches) -> {
			friends.forEach(populatedUser::addFriend);
			matches.forEach(populatedUser::addMatch);
			return populatedUser;
		});
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/handler/populator/UserPopulator.kt
package com.yahyeet.boardbook.model.handler.populator

import com.yahyeet.boardbook.model.entity.Match
import com.yahyeet.boardbook.model.entity.MatchPlayer
import com.yahyeet.boardbook.model.entity.User
import com.yahyeet.boardbook.model.repository.IMatchPlayerRepository
import com.yahyeet.boardbook.model.repository.IMatchRepository
import com.yahyeet.boardbook.model.repository.IUserRepository
import java.util.concurrent.CompletableFuture
import java.util.stream.Collectors

class UserPopulator(private val matchPlayerRepository: IMatchPlayerRepository, private val matchRepository: IMatchRepository, private val userRepository: IUserRepository) {

    fun populate(user: User): CompletableFuture<User> {
        val populatedUser = User(user.name)
        populatedUser.id = user.id

        val friendsFuture = userRepository.findFriendsByUserId(populatedUser.id)

        val futureMatchPlayers = matchPlayerRepository.findMatchPlayersByUserId(populatedUser.id)

        val futureMatches = futureMatchPlayers.thenCompose { matchPlayers ->
            val allMatchFutures = matchPlayers
                    .stream()
                    .map { matchPlayer -> matchRepository.find(matchPlayer.match!!.id) }
                    .collect<List<CompletableFuture<Match>>, Any>(Collectors.toList())

            val allOfFutureAllMatchFutures = CompletableFuture.allOf(*allMatchFutures.toTypedArray<CompletableFuture>())
            allOfFutureAllMatchFutures.thenApply { future ->
                allMatchFutures
                        .stream()
                        .map<Match>(Function<CompletableFuture<Match>, Match> { it.join() })
                        .collect<List<Match>, Any>(Collectors.toList())
            }
        }

        return friendsFuture.thenCombine(futureMatches) { friends, matches ->
            friends.forEach(Consumer<User> { populatedUser.addFriend(it) })
            matches.forEach(Consumer<Match> { populatedUser.addMatch(it) })
            populatedUser
        }
    }
}
